### PR TITLE
web: upgrade `react-router` to v6

### LINKED ISSUE #33834: React Router V6 migration
## Description

React Router V6 is out, and they have just published a [migration plan](https://github.com/remix-run/react-router/discussions/8753).

We have completed the "Setting up" stage through this PR: https://github.com/sourcegraph/sourcegraph/pull/33835

We should follow the "Migration strategy" stage to complete the migration. 

### Please follow the [migration plan](https://github.com/remix-run/react-router/discussions/8753) guide written by the React Router team themselves. I have avoided duplicating any additional information in this issue as that guide should be used as a source of truth (if it is edited).

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

STATUS: fixing one-off things based on CI checks.

- Closes #33834
- Upgraded react-router to v6
- Migrated the web application to [the data-aware router introduced in v6.4.0](https://reactrouter.com/en/main/routers/picking-a-router#using-v64-data-apis).
- Migrated `history.block` usages to the `unstable_useBlock` hook [introduced in v6.7.0](https://github.com/remix-run/react-router/issues/8139).
- Removed explicit history reference from the `renderWithBrandedContext` utility used in unit tests.
- Migrated the search-query state observer from `history.listen` to `useLocation` API.
- TODO: remove `history` from dependencies in a follow-up PR.
- TODO: prettify the `usePropt` hook borrowed from [this gist](https://gist.github.com/chaance/2f3c14ec2351a175024f62fd6ba64aa6).
- TODO: remove `repoSplat` in a follow-up PR.
- TODO: remove `historyOrNavigate` and `compatNavigate` in a follow-up PR.
- TODO: remove the top-level route in a follow-up PR.

## Test plan

CI and manually visiting all the pages 🙂

### CODE REVIEW

File: client/web/src/SearchQueryStateObserver.tsx [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+// Update search query state whenever the URL changes
+export const SearchQueryStateObserver: FC<SearchQueryStateObserverProps> = props => {
+    const { searchContextsEnabled, platformContext, setSelectedSearchContextSpec, selectedSearchContextSpec } = props
+
+    const location = useLocation()
+
+    // Execute once on mount. New values are provided in the `useEffect` hook.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    const locationSubject = useMemo(() => new BehaviorSubject<Location>(location), [])
```
Comment (philipp-spiess) [inline client/web/src/SearchQueryStateObserver.tsx:30]: I personally like to use `useState` in these situations. I remember in the very early beginnings of hooks, the React team mentioned that they have plans to not keep refs stable across renders to reduce memory usage of offscreen components or so. Don't think this was ever implemented but with a `useState` it's a bit clearer that the value is not expected to be updated when its dependencies change.

File: client/web/src/SourcegraphWebApp.tsx [hunk 2]
```diff
@@ -333,6 +296,45 @@ export class SourcegraphWebApp extends React.Component<SourcegraphWebAppProps, S
             return null
         }
 
+        const router = createBrowserRouter(
+            createRoutesFromElements(
+                <Route
+                    path="*"
```
Comment (philipp-spiess) [inline client/web/src/SourcegraphWebApp.tsx:302]: Do we still need the top level `<Route>` if we only render one element or is this just a placeholder for the future as we want to hoist more stuff up here?

File: client/web/src/repo/blob/BlameColumn.tsx [hunk 3]
```diff
@@ -64,8 +64,8 @@ export const BlameColumn = React.memo<BlameColumnProps>(
             }
 
             const subscription = codeViewElements.subscribe(codeView => {
-                if (codeView?.firstElementChild instanceof HTMLTableElement) {
-                    const table = codeView.firstElementChild
+                if (codeView?.firstElementChild) {
+                    const table = codeView.firstElementChild as HTMLTableElement
```
Comment (philipp-spiess) [inline client/web/src/repo/blob/BlameColumn.tsx:68]: This diff seems unintended?

File: client/web/src/SourcegraphWebApp.tsx (hunk 2 above)
Comment (vovakulikov) [reply client/web/src/SourcegraphWebApp.tsx:302]: I removed id in other PRs and didn't notice any problems with just Layout component as root element for the Router , should be fine to remove it

File: client/web/src/org/area/OrgHeader.tsx [hunk 4]
```diff
@@ -78,28 +74,10 @@ export const OrgHeader: React.FunctionComponent<React.PropsWithChildren<Props>>
...
+                                    ({ to, label, exact, icon: ItemIcon, condition = () => true, dynamicLabel }) =>
                                         condition(context) && (
                                             <li key={label} className="nav-item">
-                                                <NavLink
-                                                    to={url + to}
-                                                    className="nav-link"
-                                                    activeClassName="active"
-                                                    exact={exact}
-                                                    isActive={
-                                                        isActive
```
Comment (vovakulikov) [inline client/web/src/org/area/OrgHeader.tsx:98]: Is this isActive no longer needed here?

**Review by vovakulikov — APPROVED** [review]
Congrats, everyone, and especially @valerybugakov. I checked some routes manually, but just to be completely sure that we do not release 4.5 with some bad and scary regressions, I would suggest holding this PR until this Friday (release cut for the 4.5) and merging it right after, this will give us one month of in house dogfood testing, and we will be sure that 4.6 (march big release) doesn't have any big problems with routing.

File: client/web/src/org/area/OrgHeader.tsx (hunk 4 above)
Comment (valerybugakov) [reply client/web/src/org/area/OrgHeader.tsx:98]: It's implicitly calculated by [the `NavLink` component](https://reactrouter.com/en/6.8.1/components/nav-link), which uses the `active` string as a default class name for active links.

**Comment by valerybugakov** [comment]
@vovakulikov, you're right. I envision fun with merge conflicts, but we have to do this 👍

File: client/web/src/SourcegraphWebApp.tsx (hunk 2 above)
Comment (valerybugakov) [reply client/web/src/SourcegraphWebApp.tsx:302]: Yeah, it's not required. I didn't want to refactor it in this PR because it's already huge.

**Comment by valerybugakov** [comment]
Yeeees! [The main-dry-run build](https://buildkite.com/sourcegraph/sourcegraph/builds/198813) is fully green 🌪️
I'm going to work on rebasing this PR on top of this one: https://github.com/sourcegraph/sourcegraph/pull/47611

**Comment by philipp-spiess** [comment]
DO IT

**Pull request merged**